web3/encoding.nim

```diff
@@ -8,14 +8,195 @@
 # those terms.
 
 import
-  std/macros,
-  stint, ./eth_api_types, stew/[assign2, byteutils]
+  std/[sequtils],
+  faststreams/outputs,
+  stint,
+  # TODO remove assign 2 when decoding.nim is updated
+  stew/[assign2, byteutils, endians2],
+  ./eth_api_types,
+  ./abi_utils
 
-func encode*[bits: static[int]](x: StUint[bits]): seq[byte] =
-  @(x.toBytesBE())
 
-func encode*[bits: static[int]](x: StInt[bits]): seq[byte] =
-  @(x.toBytesBE())
+{.push raises: [].}
+
+type
+  AbiEncoder* = object
```

Comment:
once we have a decoder of corresponding quality, we can use https://github.com/status-im/nim-serialization?tab=readme-ov-file#defining-serialization-formats to provide the common `encode`/`decode` API

Reply:
yep, that and use https://github.com/status-im/nim-serialization/blob/2773744212584c5996131ba6e19006f66be3604a/serialization/object_serialization.nim#L27 to pick which fields to serialize (it skips the ones marked with `{.dontSerialialize.}`)

maybe `Contract` would be a better name for the format - or maybe not :)